src/CdrReader.ts

```diff
@@ -181,6 +187,73 @@ export class CdrReader {
    * Reads the member header (EMHEADER) and returns the member ID, mustUnderstand flag, and object size
    */
   emHeader(): { mustUnderstand: boolean; id: number; objectSize: number } {
+    if (this.isCDR2) {
+      return this.memberHeaderV2();
+    } else {
+      return this.memberHeaderV1();
+    }
+  }
+
+  /** XCDR1 PL_CDR encapsulation parameter header*/
+  private memberHeaderV1(): { id: number; objectSize: number; mustUnderstand: boolean } {
+    // 4-byte header with two 16-bit fields
+    this.align(4);
+    const idHeader = this.uint16();
+
+    const mustUnderstandFlag = (idHeader & 0x4000) >> 14 === 1;
+    // indicates that the parameter has a implementation-specific interpretation
+    const implementationSpecificFlag = (idHeader & 0x8000) >> 15 === 1;
+
+    // Allows the specification of large member ID and/or data length values
+    // requires the reading in of two uint32's for ID and size
+    const extendedPIDFlag = (idHeader & 0x3fff) === EXTENDED_PID;
+
+    // Indicates the end of the parameter list structure
+    const sentinelPIDFlag = (idHeader & 0x3fff) === SENTINEL_PID;
+    if (sentinelPIDFlag) {
+      throw Error("Expected Member Header but got SENTINEL_PID Flag");
+    }
+
+    // Indicates that the ID should be ignored
+    // const ignorePIDFlag = (idHeader & 0x3fff) === 0x3f03;
+
+    const usesReservedParameterId = (idHeader & 0x3fff) > SENTINEL_PID;
+
+    // Not trying to support right now if we don't need to
+    if (usesReservedParameterId || implementationSpecificFlag) {
+      throw new Error(`Unsupported parameter ID header ${idHeader.toString(16)}`);
+    }
+
+    if (extendedPIDFlag) {
+      // to do: could maybe remove because align will take care of it
+      this.uint16();
+    }
+
+    const id = extendedPIDFlag ? this.uint32() : idHeader & 0x3fff;
+    const objectSize = extendedPIDFlag ? this.uint32() : this.uint16();
+    this.originPush();
+    return { id, objectSize, mustUnderstand: mustUnderstandFlag };
+  }
+
+  private originPush(): void {
+    this.origin = this.offset;
```

Comment:
Yeah it's a bit strange the spec defines both push and pop, but doesn't use push :(

Reply:
> what does the 0 mean and why are we not using it here?
Honestly I was and am a little confused by this. I'm taking it to mean that we reset the origin to the offset because we've always been using an origin=4, which I thought came from the `PUSH(ORIGIN = 0)` that happens after writing the encHeader (encapsulation kind). If this is incorrect then we should've been failing to read 8 byte cdr messages because they could be misaligned. If it does actually set the origin to 0, then It also doesn't make sense for them to set the origin to 0 again after writing a PL_CDR header. So I've taken this to mean that it sets the origin to the offset after reading the header, so that all the offsets are relative to the end of the header, which makes a bit more sense. It seems to work for the example data that I have